### ISSUE TITLE
`cabal sdist` concatenates `main-is`

### ISSUE THREAD

**Issue by hvr** [haskell/cabal#3313]
consider the fragment

```
  if impl(ghcjs)
    main-is:           JSMain.hs
  else
    main-is:           Main.hs
```

With cabal 1.24 this causes `cabal sdist` to fail with

```
Building source dist for foo-0.1.0.0...
cabal: JSMain.hsMain.hs doesn't exist
```

@dcoutts says

>  this is a bug in the `sdist` code, it's using `flattenPackageDescription` to get both sides of all conditionals  to find all files but that doesn't work for `main-is` apparently

/cc @ezyang

**Comment by 23Skidoo** [comment]
The culprit is likely the `Monoid` instance for `Executable`, which is basically `modulePath a <> modulePath b`.

**Comment by hvr** [comment]
@23Skidoo yeah, but there's no simple fix it seems other than using a `Last` or `First` monoid instead, which would likely result in a similar effect to #3314

**Comment by 23Skidoo** [comment]
An easy fix would probably be to make it a list, and then fail at use sites if the list contains more than one element.

**Comment by dcoutts** [comment]
I think it may be better to do a dedicated traversal. We need two versions of this anyway, one to get all files under all/any configuration (for sdist) and another to find the files under a specific configuration (for build file monitoring).
